data/src/main/java/com/moez/QKSMS/repository/EmojiReactionRepositoryImpl.kt

```diff
@@ -0,0 +1,292 @@
+        }
+
+        Timber.w("No target message found for reaction text: '$originalMessageText'")
+        return null
+    }
+
+    private fun removeEmojiReaction(
+        reactionMessage: Message,
+        reaction: ParsedEmojiReaction,
+        targetMessage: Message?,
+    ) {
+        if (targetMessage == null) {
+            Timber.w("Cannot remove emoji reaction '${reaction.emoji}': no target message found")
+            return
+        }
+
+        val existingReaction = targetMessage.emojiReactions.find { candidate ->
+            candidate.senderAddress == reactionMessage.address && candidate.emoji == reaction.emoji
+        }
+
+        if (existingReaction != null) {
+            existingReaction.deleteFromRealm()
+            Timber.d("Removed emoji reaction: ${reaction.emoji} to message ${targetMessage.id}")
+        } else {
+            Timber.w("No existing emoji reaction found to remove: ${reaction.emoji} to message ${targetMessage.id}")
+        }
+
+        reactionMessage.isEmojiReaction = true
+    }
+
+    override fun saveEmojiReaction(
+        reactionMessage: Message,
+        parsedReaction: ParsedEmojiReaction,
+        targetMessage: Message?,
+        realm: Realm,
+    ) {
+        if (parsedReaction.isRemoval) {
+            removeEmojiReaction(reactionMessage, parsedReaction, targetMessage)
+            return
+        }
+
+        val reaction = EmojiReaction().apply {
+            id = keyManager.newId()
+            reactionMessageId = reactionMessage.id
+            senderAddress = reactionMessage.address
+            emoji = parsedReaction.emoji
+            originalMessageText = parsedReaction.originalMessage
+            threadId = reactionMessage.threadId
+        }
+        realm.insertOrUpdate(reaction)
+
+        if (targetMessage != null) {
+            reactionMessage.isEmojiReaction = true
+
+            // Overwrite any previous reaction from this sender for this target
+            val priorFromSender = targetMessage.emojiReactions.filter { it.senderAddress == reaction.senderAddress }
+            priorFromSender.forEach { it.deleteFromRealm() }
+
+            targetMessage.emojiReactions.add(reaction)
+
+            Timber.i("Saved emoji reaction: ${reaction.emoji} to message ${targetMessage.id}")
+        } else {
+            Timber.w("No target message, cannot save emoji reaction: ${reaction.emoji}")
+        }
+    }
+
+    override fun deleteAndReparseAllEmojiReactions(realm: Realm) {
+        val startTime = System.currentTimeMillis()
+
+        realm.delete(EmojiReaction::class.java)
+        realm.where(Message::class.java).findAll().map {
+            it.isEmojiReaction = false
+        }
+
+        val allMessages = realm.where(Message::class.java)
+            .beginGroup()
+                .beginGroup()
+                    .equalTo("type", "sms")
+                    .isNotEmpty("body")
+                .endGroup()
+                .or()
+                 .beginGroup()
```

Comment:
Sorry, it added a white-space here, on last suggestion.

```suggestion
                .beginGroup()
```